### DISCUSSION TITLE
docs don't show nested enum attribute for body

### DISCUSSION THREAD

**Discussion by Casyfill** [fastapi/fastapi#8151] [Questions] [answered]
**Describe the bug**
I have a post method with the body as pydantic object, with one attribute as Enum object; All validation works perfectly; However, two independent (?) issues occur:

1. complaint.dict() generates a dictionary - but does not flatten nested enum object into string. I am getting this is beyound the idea of `dict()`, but would be nice to have something that flattens the whole entity;
2. `complaint_type` attribute does not show up on the swagger

**To Reproduce**
example coe snippet:
```
class ComplaintType(Enum):
    other = 'other'
    commercial = "commercial"
    house = "house"
    park = "park"
    residential = "residential"
    street = "street"
    vehicle = "vehicle"


class Complaint(BaseModel):
    complaint_type:ComplaintType
    timestamp:datetime = datetime.now()
    lat:float
    lon:float
    description:str
```
**Expected behavior**
I expected
1. complaint_type to be on a swagger and in the schema as (I guess) "oneOf" element
2. be able to flatten Pydantic object into a fully json-able dictionary (converting enum objects into strings)

**Environment:**
 - OS: macOS
 - FastAPI Version  0.30.0
- Python version 3,7.3

**Comment by Sieboldianus** [comment]
Can you test whether this fixes it:

```python
class Geoaccuracy(str, Enum):
    unknown = "unknown"
    latlng = "latlng"
    place = "place"
    city = "city"
    country = "country"
```

This works for me and correctly shows in SwaggerDocs. In your case, you've specified type `Enum` but not which type for Enum (looks like its `str`, too).

**Comment by Casyfill** [comment]
thanks! that did the trick with the docs!.

I realize the other question is rather for Pydantic repo, but how should I automatically convert all nested enum values to corresponding simple values, instead? I see FastAPI does that, but wasn't able to track the exact code line

**Comment by tiangolo** [comment]
Thanks @Sieboldianus for your help!

@Casyfill it's actually an issue in Swagger UI, by adding the `str`, the generated JSON Schema also includes the type `"string"`, and Swagger UI requires it to be able to display it.

It's actually a [bug in Swagger UI](https://github.com/swagger-api/swagger-ui/issues/3761), Pydantic and FastAPI are generating a valid OpenAPI schema.

But I should update the Enum examples in the docs.

**Comment by tiangolo** [comment]
Update, I just updated the docs, to use `Enum`s that inherit from `str`: https://fastapi.tiangolo.com/tutorial/path-params/#working-with-python-enumerations

**Comment by Casyfill** [comment]
Thank you both! That makes a lot of sense - somehow I never used Enum before, so - good to know!

**Comment by tiangolo** [comment]
Thanks @Casyfill for reporting back and closing the issue! :cake:

**Comment by YuriiMotov** [comment, marked as answer]
Just `Enum` works fine in current version (FastAPI 0.116.0, Pydantic 2.11.7):

```python
from enum import Enum

from fastapi import FastAPI
from pydantic import BaseModel


class ComplaintType(Enum):
    other = "other"
    commercial = "commercial"


class Complaint(BaseModel):
    complaint_type: ComplaintType
    description: str


app = FastAPI()


@app.post("/")
def p(c: Complaint) -> Complaint:
    pass
```

![image](https://github.com/user-attachments/assets/084e6b3c-a638-42f9-90d8-8226c3008120)


In openapi.json in both cases it's:
```json
      "ComplaintType": {
        "type": "string",
        "enum": [
          "other",
          "commercial"
        ],
        "title": "ComplaintType"
      },
```
